### PR TITLE
[Android] Set Page background color

### LINKED ISSUE #5057: ContentPage BackgroundColor ignored in Forms 3.5.0.94564-pre3
### Description

ContentPage BackgroundColor is ignored in Forms 3.5.0.94564-pre3 causing the colour of a Xamarin.Android activity to show through (MainActivity or a splash activity). This is happening on device and emulator.

### Steps to Reproduce

1. Create a new Forms app.
2. Upgrade Forms nuget to v3.5-pre3. 
3. Upgrade Android.Support nugets to 28.0.0.1. (probably not necessary)
3. Set the BackgroundColor in MainPage.xaml to a non-white colour.
4. Run the app.

### Expected Behavior

MainPage uses the background colour specified in its XAML.

### Actual Behavior

MainPage has a white background (the colour of MainActivity).

### Basic Information

- Version with issue: 3.5.0.94564-pre3
- Last known good version: 3.5.0.62956-pre2
- IDE: VS 2017 v15.9.5
- Platform Target Frameworks:
  - Android: 8.1 + 9.0 (both tested)
- Android Support Library Version: 28.0.0.1 (also seen with 28.0.0)
- Affected Devices: Galaxy S6 Edge (Android 7.0) + Android Emulator

### Screenshots

Expected Behaviour: 
expectedbehaviour-35pre3

Actual Behaviour: 
actualbehaviour-35pre2

### Reproduction Link

[irongut/35pre3BGColor](https://github.com/irongut/35pre3BGColor)

### LINKED ISSUE #5073: Black page when no ShellContent.Content.BackgroundColor
### Description
When no background is set on the ShellContent.Content, the below screenshot happens.
### Steps to Reproduce

1. Make a default Shell app, not setting the background anywhere.

### Expected Behavior
A default Shell experience with a black or white page with no after effects with flyout.

### Actual Behavior
A black screen, reminiscent of yee olden XP.
### Basic Information

- Version with issue:
- Last known good version: This is the first time I've tried a Shell app
- IDE: VS 2017
- Platform Target Frameworks:
  - Android: 9, 8.1
- Nuget Packages: Xamarin.Forms: 4.0.0.94569-pre3
- Affected Devices: All emulators I've tried

### Screenshots

After closing flyout, which did look fine.
image

### PR DESCRIPTION
**Pull request by andreinitescu**
### Description of Change ###

On Android, setting Page BackgroundColor doesn't work.
Also, iOS and UWP implement a priority of BackgroundImage over BackgroundColor. If both are set BackkgroundImage is used.

### Issues Resolved ### 

* Fixes #5057 
* closes #4916
* Fixes #5073

Video with the fix (click to view full screen):

pagebkgnd

### API Changes ###

 
 None

### Platforms Affected ### 

- Android

### PR Checklist ###

- [ ] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Comment by samhouts** [comment]
@andreinitescu Can you rebase to 3.5.0 branch? Thanks!

**Comment by andreinitescu** [comment]
@samhouts Done!

**Comment by andreinitescu** [comment]
oh wait, you said 3.5.0

**Comment by samhouts** [comment]
build --uitests

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue5057.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.Collections.Generic;
+using Xamarin.Forms.CustomAttributes;
+using Xamarin.Forms.Internals;
+using Xamarin.Forms.Xaml;
+
+namespace Xamarin.Forms.Controls.Issues
+{
+	[Preserve(AllMembers = true)]
```
Comment (samhouts) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue5057.xaml.cs]:
```suggestion
#if APP
	[Preserve(AllMembers = true)]
```

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue5057.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+		void BtnSetBkgndColorDefault_Clicked(object sender, System.EventArgs e)
+		{
+			BackgroundColor = (Color)BackgroundColorProperty.DefaultValue;
+		}
+
+		void BtnSetBkgndImgNull_Clicked(object sender, System.EventArgs e)
+		{
+			BackgroundImage = null;
+		}
+	}
```
Comment (samhouts) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue5057.xaml.cs]:
```suggestion
	}
#endif
```

**Comment by andreinitescu** [comment]
@samhouts  Done!

**Comment by samhouts** [comment]
build --uitests

File: Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs [hunk 3]
```diff
@@ -105,24 +104,27 @@ void UpdateHeight()
...
-			if (!string.IsNullOrEmpty(view.BackgroundImage))
-				this.SetBackground(Context.GetDrawable(view.BackgroundImage));
+			if (Element.Parent is BaseShellItem)
+			{
+				Color background = Element.BackgroundColor;
+				if (background.IsDefault)
+					base.UpdateBackgroundColor();
+				else
+					SetBackgroundColor(Context.Resources.GetColor(global::Android.Resource.Color.BackgroundLight, Context.Theme));
+			}
```
Comment (mattleibow) [inline Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs]: I am not too sure about the shell bits, but I think this has changed from

```csharp
if (shell) {
  if (isDefault) {
    SetBackground(GetFromTheme());
  } else {
    SetBackground(background.ToAndroid());
  }
}
```

to 

```csharp
if (shell) {
  if (isDefault) { // this has flipped
    base.Update();
  } else {
    SetBackground(GetFromTheme());
  }
}
```

You may be wanting:

```csharp
// a special case for shell
if (view.Parent is BaseShellItem && view.BackgroundColor == Color.Default) {
  var color = Context.Resources.GetColor(global::Android.Resource.Color.BackgroundLight, Context.Theme);
  SetBackgroundColor(color);
} else {
  UpdateBackgroundColor(); // we may not need the `base.` as we no longer have a method with the same name
}

**Comment by andreinitescu** [comment]
All done.
I also added some micro optimizations, reading dependency properties once, not calling native API unless it's really necessary.

File: Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs [hunk 4]
```diff
@@ -105,24 +104,25 @@ void UpdateHeight()
...
-		{
-			if (!string.IsNullOrEmpty(view.BackgroundImage))
-				this.SetBackground(Context.GetDrawable(view.BackgroundImage));
+			string bkgndImage = page.BackgroundImage;
+			if (!string.IsNullOrEmpty(bkgndImage))
+				this.SetBackground(Context.GetDrawable(bkgndImage));
 			else
-				this.SetBackground(null);
+			{
+				Color bkgndColor = Element.BackgroundColor;
```
Comment (mattleibow) [inline Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs]:
```suggestion
				Color bkgndColor = page.BackgroundColor;
```

**Comment by andreinitescu** [comment]
@mattleibow Argh, good eye! Done!

**Comment by mattleibow** [comment]
Fixed the merge conflict... hope I didn't break it 😬

**Comment by mattleibow** [comment]
build

**Pull request merged**